### PR TITLE
Random Battles: Refactor for code standards and readability

### PR DESCRIPTION
**Pull request by AnnikaCodes**
This PR
- Brings all Random Battles code up to the standard 120-charaacter line length limit
- Improves readability for all Random Battles code
- Refactors current-gen Random Battles team generation to be more modular and readable

Ideas for the future:
- Further refactors (I'm waiting on this until I'm more familiar with the codebase)
- Possibly adding unit tests
- Ideally, properly divorcing code and data like Battle Factory has

cc @KrisXV @ACakeWearingAHat

### CODE REVIEW

File: data/mods/gen4/random-teams.ts [hunk 1]
```diff
@@ -354,59 +394,109 @@ export class RandomGen4Teams extends RandomGen5Teams {
...
+						(hasAbility['Guts'] && hasType['Normal'] && movePool.includes('facade')) ||
+						(hasAbility['Slow Start'] && movePool.includes('substitute')) ||
+						(counter['defensesetup'] && !counter.recovery && !hasMove['rest']) ||
+						(movePool.includes('spore') || (!moves.some(id => recoveryMoves.includes(id)) && (
+							movePool.includes('softboiled') ||
+							(species.baseSpecies === 'Arceus' && movePool.includes('recover'))
+						))) ||
+						(species.requiredMove && movePool.includes(toID(species.requiredMove)))
+					)
+				) {
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/mods/gen4/random-teams.ts:493]: Yeah, there is. In the current generation, I refactored each line into its own `moveRejectionChecker`. I'm planning to refactor the old generations at some point, but I want to see how the current-gen refactor goes first (I just don't like the randbats code having a ridiculously long `max-len` setting).

File: data/mods/gen3/random-teams.ts [hunk 2]
```diff
@@ -415,17 +446,25 @@ export class RandomGen3Teams extends RandomGen4Teams {
...
+			(hasMove['bellydrum'] && counter.Physical - counter['priority'] > 1) ||
+			(hasMove['swordsdance'] && counter.Status < 2)
+		) {
 			item = 'Salac Berry';
-		} else if (hasMove['endure'] || (hasMove['substitute'] && (hasMove['endeavor'] || hasMove['flail'] || hasMove['reversal']))) {
-			item = (species.baseStats.spe <= 100 && ability !== 'Speed Boost' && !counter['speedsetup'] && !hasMove['focuspunch']) ? 'Salac Berry' : 'Liechi Berry';
+		} else if (hasMove['endure'] || (hasMove['substitute'] && ['endeavor', 'flail', 'reversal'].some(m => hasMove[m]))) {
+			item = (species.baseStats.spe <= 100 && ability !== 'Speed Boost' && !counter['speedsetup'] && !hasMove['focuspunch']) ?
+				'Salac Berry' :
+				'Liechi Berry';
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/mods/gen3/random-teams.ts:457]: Yeah that's fair tbh, I'll change those around.

File: data/mods/gen5/random-teams.ts [hunk 3]
```diff
@@ -295,42 +310,87 @@ export class RandomGen5Teams extends RandomGen6Teams {
...
-				if (counter.setupType === 'Special' && moveid === 'hiddenpower' && species.types.length > 1 && counter['Special'] <= 2 && !hasType[move.type] && !counter['Physical'] && counter['specialpool']) {
+				if (
+					counter.setupType === 'Special' &&
+					moveid === 'hiddenpower' &&
+					species.types.length > 1 &&
+					counter['Special'] <= 2 &&
+					!hasType[move.type] &&
+					!counter['Physical'] &&
+					counter['specialpool']
+				) {
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/mods/gen5/random-teams.ts:326]: The rubric I used was that if a line didn't fit I broke it up, trying to group together similar conditions.

File: data/mods/gen6/random-teams.ts [hunk 4]
```diff
@@ -176,17 +180,34 @@ export class RandomGen6Teams extends RandomGen7Teams {
...
+						(counter[move.type] > 1 && counter.Status > 1) ||
+						(hasAbility['Sheer Force'] && !!counter['sheerforce'])
+					) rejected = true;
 					break;
 				case 'defog':
-					if (counter.setupType || hasMove['spikes'] || hasMove['stealthrock'] || (hasMove['rest'] && hasMove['sleeptalk']) || teamDetails.defog) rejected = true;
+					if (
+						counter.setupType ||
+						hasMove['spikes'] || hasMove['stealthrock'] ||
+						(restTalk) ||
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/mods/gen6/random-teams.ts:196]: Yeah, that's on me. I'll fix it.

File: data/random-teams.ts [hunk 5]
```diff
@@ -82,28 +227,12 @@ export class RandomTeams {
...
-	// hasMegaEvo(species) {
-	// 	if (!species.otherFormes) return false;
-	// 	const firstForme = this.dex.getSpecies(species.otherFormes[0]);
-	// 	return !!firstForme.isMega;
-	// }
+	isBeneficialInSingles(move: Move) {
+		return [
+			'facade', 'lightscreen', 'reflect', 'sleeptalk', 'spore', 'substitute', 'switcheroo', 'teleport', 'toxic', 'trick',
+		].includes(move.id) && (move.category !== 'Status' || !move.flags.heal);
+	}
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/random-teams.ts:234]: I'm not entirely sure what the point of this either; it's just extracting https://github.com/smogon/pokemon-showdown/blob/master/data/random-teams.ts#L997 out into a method. It should be moves that _aren't_ in the Array literal, though, so I fixed that and I gave it a less pretentious name.

File: data/random-teams.ts [hunk 6]
```diff
@@ -181,7 +315,12 @@ export class RandomTeams {
 			if (pool.length <= 4) {
 				moves = pool;
 			} else {
-				moves = [this.sampleNoReplace(pool), this.sampleNoReplace(pool), this.sampleNoReplace(pool), this.sampleNoReplace(pool)];
+				moves = [
+					this.sampleNoReplace(pool),
+					this.sampleNoReplace(pool),
+					this.sampleNoReplace(pool),
+					this.sampleNoReplace(pool),
+				];
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/random-teams.ts:323]: Sounds like a good idea!

File: data/random-teams.ts [hunk 7]
```diff
@@ -469,88 +592,91 @@ export class RandomTeams {
...
 			}
 			// Moves with low accuracy:
-			if (move.accuracy && move.accuracy !== true && move.accuracy < 90) counter['inaccurate']++;
+			if (move.accuracy && move.accuracy !== true && move.accuracy < 90) counter.inaccurate++;
 
 			// Moves that change stats:
-			if (RecoveryMove.includes(moveid)) counter['recovery']++;
-			if (ContraryMove.includes(moveid)) counter['contrary']++;
+			if (RecoveryMove.includes(moveid)) counter.recovery++;
+			if (StatDroppingMoves.includes(moveid)) counter.contrary++;
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/random-teams.ts:651]: That's fine; my preference for it was quite minimal and I'm happy to change it back.

File: data/mods/gen7/random-teams.ts [hunk 8]
```diff
@@ -491,50 +536,118 @@ export class RandomGen7Teams extends RandomTeams {
...
+						!counter['Dark'] &&
+						!hasAbility['Protean']
+					) || (
+						hasMove['suckerpunch'] && !hasAbility['Contrary'] && counter.stab < species.types.length
+					) || (
+						hasType['Dragon'] &&
+						!counter['Dragon'] &&
+						!hasAbility['Aerilate'] && !hasAbility['Pixilate'] &&
+						!hasMove['fly'] && !hasMove['rest'] && !hasMove['sleeptalk']
+					) || (hasType['Electric'] && (!counter['Electric'] || movePool.includes('thunder'))) ||
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/mods/gen7/random-teams.ts:589]: I'm planning to get rid of the gargantuan `if` statements altogether; this is just a temporary stopgap measure so we can stop making exceptions with `eslint`.

File: data/random-teams.ts [hunk 9]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+			return {cull: hasMove['uturn'] && !counter.setupType};
+		case 'leechlife':
+			return {cull: (isDoubles && hasMove['lunge']) || movePool.includes('firstimpression') || movePool.includes('spikes')};
+			break;
+		case 'stoneedge':
+			const gutsCullCondition = hasAbility['Guts'] && (!hasMove['dynamicpunch'] || hasMove['spikes']);
+			const rockSlidePlusStatusPossible = counter.Status && movePool.includes('rockslide');
+			return {cull: gutsCullCondition || rockSlidePlusStatusPossible || hasMove['rockblast'] || hasMove['rockslide']};
+		case 'poltergeist':
+			return {cull: hasMove['knockoff']};
```
Comment (ACakeWearingAHat) [inline data/random-teams.ts:962]:
```suggestion
			// Special case for Dhelmise in Doubles, which doesn't want both
			return {cull: hasMove['knockoff']};
```
honestly just remove this and remove poltergeist from dhelm and that'd be fine too. simpler, not much impact. poltergeist sucks in doubles anyway

File: data/random-teams.ts [hunk 10]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+			return {cull: hasType['Steel'] && hasMove['shadowsneak'] && counter.Physical < 4};
+		case 'dragonpulse': case 'spacialrend':
+			return {cull: hasMove['dracometeor'] && counter.Special < 4};
+		case 'darkpulse':
+			const defogger = hasMove['defog'] && counter.setupType !== 'Special';
+			return {cull: ['foulplay', 'knockoff', 'suckerpunch'].some(m => hasMove[m]) || defogger};
+		case 'suckerpunch':
+			return {cull: hasMove['rest'] || counter.damagingMoves.length < 2 || (counter.Dark > 1 && !hasType['Dark'])};
+		case 'meteormash':
+			return {cull: movePool.includes('extremespeed')};
```
Comment (ACakeWearingAHat) [inline data/random-teams.ts:980]:
```suggestion
			// Special case for Lucario, which always wants Extreme Speed.
			return {cull: movePool.includes('extremespeed')};
```
rejected move should probably be ice punch and not this but that's a discussion for later

File: data/random-teams.ts [hunk 11]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+			return {cull: ['fleurcannon', 'moonblast', 'petaldance'].some(m => hasMove[m])};
+
+		// Status:
+		case 'bodyslam': case 'clearsmog':
+			const toxicCullCondition = hasMove['toxic'] && !hasType['Normal'];
+			return {cull: hasMove['sludgebomb'] || hasMove['trick'] || movePool.includes('recover') || toxicCullCondition};
+		case 'haze':
+			return {cull: !teamDetails.stealthRock && (hasMove['stealthrock'] || movePool.includes('stealthrock'))};
+		case 'hypnosis':
+			return {cull: hasMove['voltswitch']};
```
Comment (ACakeWearingAHat) [inline data/random-teams.ts:991]:
```suggestion
			// Special case for Xurkitree to properly split Blunder Policy and Choice item sets
			return {cull: hasMove['voltswitch']};
```

File: data/random-teams.ts [hunk 12]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+			return {cull: movePool.includes('extremespeed')};
+		case 'dazzlinggleam':
+			return {cull: ['fleurcannon', 'moonblast', 'petaldance'].some(m => hasMove[m])};
+
+		// Status:
+		case 'bodyslam': case 'clearsmog':
+			const toxicCullCondition = hasMove['toxic'] && !hasType['Normal'];
+			return {cull: hasMove['sludgebomb'] || hasMove['trick'] || movePool.includes('recover') || toxicCullCondition};
+		case 'haze':
+			return {cull: !teamDetails.stealthRock && (hasMove['stealthrock'] || movePool.includes('stealthrock'))};
```
Comment (ACakeWearingAHat) [inline data/random-teams.ts:989]:
```suggestion
			// Special case for Corsola-Galar, which always wants Will-O-Wisp
			return {cull: !teamDetails.stealthRock && (hasMove['stealthrock'] || movePool.includes('stealthrock'))};
```

File: data/random-teams.ts [hunk 13]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+		case 'hypnosis':
+			return {cull: hasMove['voltswitch']};
+		case 'willowisp': case 'yawn':
+			return {cull: hasMove['thunderwave'] || hasMove['toxic']};
+		case 'painsplit': case 'recover': case 'synthesis':
+			return {cull: hasMove['rest'] || hasMove['wish'] || (move.id === 'synthesis' && hasMove['gigadrain'])};
+		case 'roost':
+			return {cull: hasMove['throatchop'] || (hasMove['stoneedge'] && !hasType['Rock'])};
```
Comment (ACakeWearingAHat) [inline data/random-teams.ts:997]:
```suggestion
			// Special case for Hawlucha, which doesn't want Roost + 3 attacks
			return {cull: hasMove['throatchop'] || (hasMove['stoneedge'] && !hasType['Rock'])};
```

File: data/random-teams.ts [hunk 14]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+		case 'reflect': case 'lightscreen':
+			return {cull: !!teamDetails.screens};
+		case 'substitute':
+			const moveBasedCull = ['bulkup', 'painsplit', 'roost'].some(m => movePool.includes(m));
+			const doublesPowerWhip = isDoubles && movePool.includes('powerwhip');
+			const calmMindCullCondition = !counter.recovery && movePool.includes('calmmind');
+			return {cull: hasMove['rest'] || moveBasedCull || doublesPowerWhip || calmMindCullCondition};
+		case 'helpinghand':
+			return {cull: hasMove['acupressure']};
```
Comment (ACakeWearingAHat) [inline data/random-teams.ts:1006]:
```suggestion
			// Special case for Shuckle in Doubles, which doesn't want sets with no method to harm foes
			return {cull: hasMove['acupressure']};
```

**Review by ACakeWearingAHat — CHANGES_REQUESTED** [review]
Add comments specifying hardcoded move-cull lines and their reasonings

File: data/random-teams.ts [hunk 15]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+			return {cull: counter.speedsetup || (hasType['Rock'] && counter.Status) || uturnCullCondition};
+		case 'blazekick':
+			return {cull: counter.Special >= 1};
+		case 'firefang': case 'flamethrower':
+			const otherFireMoves = ['heatwave', 'overheat'].some(m => hasMove[m]);
+			return {cull: (hasMove['fireblast'] && counter.setupType !== 'Physical') || otherFireMoves};
+		case 'overheat':
+			return {cull: hasMove['flareblitz'] || (isDoubles && hasMove['calmmind'])};
+		case 'aquajet': case 'psychicfangs':
+			return {cull: hasMove['rapidspin'] || hasMove['taunt']};
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/random-teams.ts:875]: I would like this PR to not change functionality, but we can talk about how best to change this - Morpeko could be directly hardcoded.

File: data/random-teams.ts [hunk 16]
```diff
@@ -559,22 +685,740 @@ export class RandomTeams {
...
+		case 'aquatail': case 'flipturn': case 'retaliate':
+			return {cull: hasMove['aquajet'] || counter.Status};
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/random-teams.ts:877]: See my comment re: Morpeko

**Comment by Zarel** [comment]
<3

**Pull request merged**